### PR TITLE
Document round-off error in `.mod_euc()`-method, see issue #50179

### PR DESCRIPTION
**Pull request by UnderTheCarpet**
Due to a round-off error the method `.mod_euc()` of both `f32` and `f64` can produce mathematical invalid outputs. If `self` in magnitude is much small than the modulus `rhs` and negative, `self + rhs` in the first branch cannot be represented in the given precision and results into `rhs`. In the mathematical strict sense, this breaks the definition. But given the limitation of floating point arithmetic it can be thought of the closest representable value to the true result, although it is not strictly in the domain `[0.0, rhs)` of the function. It is rather the left side asymptotical limit. It would be desirable that it produces the mathematical more sound approximation of `0.0`, the right side asymptotical limit. But this breaks the property, that `self == self.div_euc(rhs) * rhs + a.mod_euc(rhs)`.

The discussion in issue #50179 did not find an satisfying conclusion to which property is deemed more important. But at least we can document the behaviour. Which this pull request does.

### CODE REVIEW

File: src/libstd/f32.rs [hunk 1]
```diff
@@ -254,6 +257,8 @@ impl f32 {
     /// assert_eq!((-a).mod_euc(b), 1.0);
     /// assert_eq!(a.mod_euc(-b), 3.0);
     /// assert_eq!((-a).mod_euc(-b), 1.0);
+    /// // limitation due to round-off error
+    /// assert!((-std::f32::EPSILON).mod_euc(3.0) != 0.0);
```
[the remark this answers is not in the document]
Comment (fkjogu) [reply src/libstd/f32.rs]: Floating point arithmetic is many things, but it is deterministic :wink:. I could turn this assertion around, but I put it that way ('not equal to') to stress, that it results in something one might not expect. The way proposed by you reads more like expected behaviour.

The [RFC](https://github.com/rust-lang/rfcs/pull/2169) did not talk about border cases of floats. I could also add these, if you'd like that:

```rust
    assert!(std::f64::INFINITY.mod_euc(a).is_nan());
    assert_eq!(a.mod_euc(std::f64::INFINITY), a);
    assert!(a.mod_euc(std::f64::NAN).is_nan());
    assert!(std::f64::INFINITY.mod_euc(std::f64::INFINITY).is_nan());
    assert!(std::f64::INFINITY.mod_euc(std::f64::NAN).is_nan());
    assert!(std::f64::NAN.mod_euc(std::f64::INFINITY).is_nan());
```
which all are true as expected. In particular `assert_eq!(a.mod_euc(std::f64::INFINITY), a);` is nice. The rest might be a bit noisy?

[the remark this answers is not in the document]
Comment (scottmcm) [reply src/libstd/f32.rs]: > The way proposed by you reads more like expected behaviour.

I don't know what's best.  I kind-of like not calling it _error_, but putting a positive spin on it and saying "yes, this is the closest float to the mathematically-correct value".

Whatever you choose to do, as someone from the "but I thought the whole point was to have a half-open range" camp, I'd like to see enough of the "why this behaviour isn't a bug" discussion in the docs to keep someone from me from opening another issue about it 😆 

> The rest might be a bit noisy?

Maybe pick some that are particularly interesting for the examples section, describe the usual things like "any input being NaN gives NaN" in prose, and add (non-doc-)tests for the rest?  The only such tests I could find for `mod_euc` so far is this:

https://github.com/rust-lang/rust/blob/52ed3d8761dc42e51429eb3bbf56a94d14fdd5c6/src/libcore/tests/num/int_macros.rs#L33-L36

**Comment by shepmaster** [comment]
Ping from triage, @withoutboats ! This is awaiting your review.

**Comment by pietroalbini** [comment]
Ping from triage! This PR needs a review, can @withoutboats or someone else from @rust-lang/libs review this?

**Comment by alexcrichton** [comment]
@fkjogu did you want to address @scottmcm's [comment](https://github.com/rust-lang/rust/pull/50342#discussion_r185164446) or is this good to go? The tweaks look fine by me (naively!)

**Comment by fkjogu** [comment]
I'm going to add the unit tests for the special float values (`INF`, `NAN`). i do not think it is necessary to spell it out explicitly: it is expected behavior overall.

But I find it hard to include a rational on choosing the given drawbacks in the documentation, because I'm not fully convinced, this is the way to go for this method.

The reason I haven't done the above already is that ideally I'd like to proof that `self == self.div_euc(rhs) * rhs + a.mod_euc(rhs)` is always true with the given implementation before adding this property to the documentation as well. I'll update soon.

**Comment by TimNN** [comment]
Ping from triage, @fkjogu: could you give us an update on your plans for this PR?

**Comment by alexcrichton** [comment]
Looks like there may be some travis errors to help clean up?

**Comment by fkjogu** [comment]
I blindly added the tests to `libcore/tests/num/int_macros.rs`, but now I realize it is just the wrong spot. What is the correct location to add unit tests for methods on `f32/f64`?

**Comment by alexcrichton** [comment]
Ah anywhere within `libcore/tests` should be fine

**Comment by fkjogu** [comment]
I've moved the unit tests to `libcore/tests/mod.rs`. When I run the tests, it works for `.mod_euc()` but for `.div_euc()` I get
```
no method named `div_euc` found for type `{float}` in the current scope
```
and I do not understand why. Sorry for the trouble.

**Comment by shepmaster** [comment]
> and I do not understand why.

A floating point literal like `42.0` doesn't have a fixed type yet; it can be `f32` or `f64` depending on context.

The difference is that the first test uses a known-type value (e.g. INFINITY) first. Since there's only the one implementation of the method when the receiver type is known, type inference can pin down the type of `a`. However, the second test uses the literal first, but there's multiple implementations for the method (depending on if it's `f32` or `f64`):

```rust
#![feature(euclidean_division)]

use std::f32::INFINITY;

#[test]
fn mod_euc() {
    let a = 42.0;
    assert!(INFINITY.mod_euc(a).is_nan());
}

#[test]
fn div_euc() {
    let a = 42.0;
    assert_eq!(a.div_euc(INFINITY), 0.0);
}
```

TL;DR: pin down the type:

```rust
let a: $fty = 42.0;
```

**Comment by fkjogu** [comment]
Thanks, that solved the issue. Good to go.

**Comment by pietroalbini** [comment]
The reviewer is away, maybe someone else from @rust-lang/libs can review this?

**Comment by pietroalbini** [comment]
The reviewer is away, maybe someone else from @rust-lang/libs can review this?

**Comment by BurntSushi** [comment]
The diff LGTM and it looks like other folks who know more about this topic than I do are happy as well, so @bors r+

(@Mark-Simulacrum FYI, I removed myself from the rust-highfive rotation because my time is so constrained. Happy to do the occasional review though!)

**Comment by pietroalbini** [comment]
@fkjogu you need to rebase on the latest master, even if you don't see any conflict (sorry about that!)

**Comment by fkjogu** [comment]
I've rebased (a few hours ago). How can I check, if I need to rebase again?

**Comment by pietroalbini** [comment]
That should be enough, thanks!

@bors r=BurntSushi rollup
